README.md
# We Were Soldiers Once... and Young

We will be landing under fire gentlemen. Men will die. When Crazy Horse was a baby, he nursed at the breast of every woman in the tribe. The Sioux raised their children that way. Every warrior called every woman in the tribe ‘Mother.’ Every older warrior, they called him ‘Grandfather.’ Now, the point here is that they fought as a family. Take care of your men. Teach them to take care of each other. ‘Cause when this starts… each other is all we’re gonna have.

Let us pray.

Our Father in Heaven, before we go into battle, every soldier among us will approach you each in his own way. Our enemies too, according to their own understanding, will ask for protection and for victory. And so, we bow before your infinite wisdom. We offer our prayers as best we can. I pray you watch over the men that I lead into battle. You use me as your instrument in this awful hell of war to watch over them. Especially if they’re men like these before me, deserving of a future in your blessing and goodwill. Amen.

Oh, yes, and one more thing, dear Lord, about our enemies, ignore their heathen prayers and help us blow those little b***ards straight to Hell. Amen.”

1. Lift at Firebase Falcon, time 1035. Slicks embark troopers of 1st and 2nd battalions, 7th Cavalry. Head due west towards the Ia Drang valley, in the central highlands east of the Chu Pong Massif.
2. Fly nap of the earth, using the dry creek bed for cover and follow the road towards the west end of the valley. Spec Ops say there are Viet Cong and NVA units with Chinese–made RPGs and 23mm antiaircraft cannons. Expect Charlie to patrol the area.
3. Deploy troopers in the clearing in front of the roadblock, LZ X–Ray. Direct the elements from the 21st Field Artillery regiment at Firebase Falcon to begin bombardment. Use artillery wisely.
4. Slicks return to Firebase Falcon for more troopers. Gunships support the assault with mini–guns and white phosphorus rockets.
5. This is a war of attrition. The side with the highest enemy body count wins. Each player gets 1 point per hit and scores unit hit points for each kill; but looses 50 points for a crash or 10 points for an ejection. Slick pilots receive triple points for every hit by Cavalry units that they disembark!
6. Refuel and rearm when rotors stop. Re–slotting deducts 25 points from your personal score.
7. Pick up and drop off chalks as many times as you like, even in the field. It happens automatically when you land and open the cockpit door within 100 metres of  a chalk. Disembark troopers the same way: land and open the door.
8. Calm your nerves by switching on the FM radio tuned to Armed Forces Radio Vietnam, transmitting out of Saigon on 30.00FM.
9. If pilots available, F-5E-3 Skyraider and F-86F Checkertail flights out of Saigon can also support the assault, but beware: Chinese Shenyang J-5s of the Vietnam People's Air Force 921 fighter regiment, the so-called Red Star squadron, may be patrolling the Chu Pong sector.

Are you wondering what was going through Custer’s mind when he realised that he’d led his men into a slaughter? Don’t. Custer was a pussy. You ain’t. You need us, you call us!

Play as Lt Col Nguyen Huu An, commanding the Vietnamese People’s Army from a bunker on the slopes of Chu Pong. Move inside the column, grab them by the belt, and thus avoid casualties from the artillery and air.


CHANGELOG.md
# Change Log

## Version 1.0, 6 September 2020

- No mini-guns on the slicks
- Fix parking places
- 3,000-strong VPA division
- 6am start
- Early dawn, more fog, more aggression
- Updated description, including TL;DR
- Match blue to red buls
- SetFrequency and TransmitMessage for 31 FM beacon
- Score reports Cav and VPA maximums
- LZ X-ray changed to EJ83; groundTurbulence increased
- Blue/Red airbases changed to Gudauta and Maykop
- Light fog
- Airbase alignments
- New random dynamic weather
- Enable dust
- Month of June
- Fixed-wing start in air, pre-equiped; (x, y) location adjustments for Caucasus
  terrain; warehouses updated
- Switch to Caucasus theatre; LM30
- Remove redundant array-table indices
- Elide redundant difficulty options

Lots of changes. Primarily, relocated from Nevada to the Caucasus map. Belated
release. Thanks to NoFear1979 for prompting an update.

## Version 0.7, 21 September 2018

- Upgraded for DCS v2.5.3.21708
- Randomise vpa.attackInZone task assigns
- Avoid lagging
- Labels are NONE, i.e. 0
- Upgrade R3DL1N3 to 98863b5

Many thanks to Durham and the [16th Air Guards Regiment](http://www.16agr.com)
for their feedback and testing.

## Version 0.6, 18 August 2017

- Merge branch 'feature/2.1.1.8491' into develop
- Group clean-up every 30s; stagger spawning and attacking triggers
- Guard for empty groups
- VPA attack off-road at 10m/s
- Updated R3DL1N3 script
- Adjusted options
- New 2.1.1 Update 2 options
- Task selected
- Need to require modules

## Version 0.5, 11 May 2017

- Added note about triple scoring for disembarked chalks
- Fix rotation: heading increases anti-clockwise

## Version 0.4, 12 April 2017

- Three extra AAA units, spawned when Skyraiders come alive
- Added three-element Checkertail flight
- Added F-86F Checkertails to briefing
- Pretend the MiG-15s are Chinese Shenyang J-5s

## Version 0.3, 11 April 2017

- Creech AFB is red, Nellis AFB is blue
- Added Skyraider and Redstar flights
- Play music on 30.00FM
- Destroy empty groups every five minutes
- Added Group.destroyAllEmpty function
- Fix embedded quotes: single not double
- Always play a tune every five minutes; players can turn it off now
- Guard against events with no initiator
- Briefing about Armed Forces Radio Vietnam
- Fix chalk embarking-disembarking briefing
- Guard conditions for events with no initiator added
- Resampled audio files to 22.05kHz; no normalisation

## Version 0.2, 3 April 2017

- Nil’ify chalk when player enters unit
- Report KIA every 150 seconds
- Sort scores by score key; plus fixes
- Output scores at least every 150 seconds
- Record friendly and hostile hits and kills
- Kills score the target’s initial hit points
- Units known who disembarked them
- Load and unload troopers via the door

## Version 0.1, 27 March 2017

Two factions. Reds are Vietnam, China and Russia. Blues are USA. It's 14
November 1965. The day begins at 1035 hours. It's 22 degrees Celcius with a
light wind. Mount Charleston is now the Ia Drang valley, Vietnam. Red bulls
marks LZ X-Ray; blue bulls marks Firebase Falcon, about ten miles to the west.
Firebase Falcon is a grass airfield. Note, grass airfield fueling and re-arming
is a broken feature at version 2.0.x of DCS.


l10n/DEFAULT/We Were Soldiers Once.lua
-- R3DL1N3/We Were Soldiers Once.lua
--
-- Copyright © 2017, 2018, R3DL1N3, United Kingdom
--
-- Permission is hereby granted, free of charge, to any person obtaining a copy
-- of this software and associated documentation files (the “Software”), to
-- deal in the Software without restriction, including without limitation the
-- rights to use, copy, modify, merge, publish, distribute, sublicense, and/or
-- sell copies of the Software, and to permit persons to whom the Software is
-- furnished to do so, subject to the following conditions:
--
--   The above copyright notice and this permission notice shall be included in
--   all copies or substantial portions of the Software.
--
-- THE SOFTWARE IS PROVIDED “AS IS,” WITHOUT WARRANTY OF ANY KIND, EITHER
-- EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
-- MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NON-INFRINGEMENT. IN NO
-- EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES
-- OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
-- ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
-- DEALINGS IN THE SOFTWARE.
--
--------------------------------------------------------------------------------

function Unit:addScore(key, score)
  self:addPlayerScore(key, 1)
  self:addPlayerScore('Score', score)
end

-- Periodically outputs player scores to all sides.
world.addEventFunction(function(event)
  if event.id == world.event.S_EVENT_PLAYER_SCORED then
    UserFlag['77'] = (UserFlag['77'] or 0) + 1
  end
end)

slick = {}

-- Slicks can embark all troopers, including tasked troopers.
function slick.embarkOrDisembark(unit)
  if not unit then return end
  if not unit:isLanded() then return end
  local groups = table.fromiter(Group.filtered(coalition.side.BLUE, Group.Category.GROUND, function(group)
    return group:getSize() > 0 and cav.names:includesGroup(group)
  end))
  unit:embarkOrDisembark(groups, 100, 100)
end

world.addEventFunction(function(event)
  -- Sometimes there is no initiator. Just ignore it in that case.
  if not event.initiator then return end
  if event.id == world.event.S_EVENT_CRASH then
    trigger.action.outText(event.initiator:getName() .. ' crashed', 3)
    event.initiator:addScore('crashes', -50)
  elseif event.id == world.event.S_EVENT_EMBARKED then
    local text = event.initiator:getName() .. ' embarked ' .. event.units.name
    trigger.action.outTextForCoalition(coalition.side.BLUE, text, 3)
  elseif event.id == world.event.S_EVENT_DISEMBARKED then
    local text = event.initiator:getName() .. ' disembarked ' .. event.units.name
    trigger.action.outTextForCoalition(coalition.side.BLUE, text, 3)
  elseif event.id == world.event.S_EVENT_EJECTION then
    trigger.action.outText(event.initiator:getName() .. ' ejected', 3)
    event.initiator:addScore('ejects', -10)
  elseif event.id == world.event.S_EVENT_PLAYER_LEAVE_UNIT then
    event.initiator:addScore('leaves', -25)
  elseif event.id == world.event.S_EVENT_HIT then
    local score
    if event.initiator:isHostileWith(event.target) then
      score = event.initiator:disembarkedBy() and 3 or 1
    elseif event.initiator:isFriendlyWith(event.target) then
      score = -10
    else
      score = 0
    end
    event.initiator:addScore('hits', score)
    if event.target:getLife() < 1 then
      local life = event.target.getLife0 and event.target:getLife0() or 1
      event.initiator:addScore('kills', life * score)
    end
  end
end)

--------------------------------------------------------------------------------
--                                                                           kia
--------------------------------------------------------------------------------

kia = {}

world.addEventFunction(function(event)
  if not event.initiator then return end
  if event.id == world.event.S_EVENT_DEAD then
    local unit = event.initiator
    if not unit.getCoalition then return end
    local side = unit:getCoalition()
    kia[side] = (kia[side] or 0) + 1
    UserFlag['7'] = (UserFlag['7'] or 0) + 1
  end
end)

local cavMax = 1000
local vpaMax = 3000

function kia.out(seconds)
  local cav = kia[coalition.side.BLUE] or 0
  local vpa = kia[coalition.side.RED] or 0
  local text = 'KIA Cav ' .. cav .. '/' .. cavMax .. ', VPA ' .. vpa .. '/' .. vpaMax
  trigger.action.outText(text, seconds or 3)
end

--------------------------------------------------------------------------------
--                                                                           cav
--------------------------------------------------------------------------------

cav = {
  zone = Zone('Cav'),
  -- Names for chalks and troopers.
  names = Names('Cav Chalk #', 'Cav Trooper #'),
}

-- Spawns one chalk of Cavalry.
function cav.spawnChalk(fromZone, toZone)
  if #Unit.allInZone(fromZone, coalition.side.BLUE, Group.Category.GROUND) ~= 0 then return end
  if cav.names.unitCounter.counter >= cavMax then return end
  if #table.fromiter(Unit.filtered(coalition.side.BLUE, Group.Category.GROUND)) >= 150 then return end
  local units = Units()
  units:addType('Soldier M4', 7)
  units:addType('Soldier M249', 1)
  units:setExcellentSkill()
  units:setRandomTransportable(false)
  cav.names:applyTo(units)
  units:formSquareInZones(fromZone, cav.zone)
  if toZone then
    local mission = Mission()
    local unit = units:all()[1]
    mission:addTurningPoint(unit.x, unit.y)
    mission:turningToPoint(toZone.point)
    units.route = {points = mission:waypoints()}
  end
  units:spawn(country.id.USA, Group.Category.GROUND)
end

--------------------------------------------------------------------------------
--                                                                           vpa
--------------------------------------------------------------------------------

vpa = {
  zone = Zone('VPA'),
  names = Names('VPA Squad #', 'VPA Soldier #'),
}

function vpa.spawnSquad(zone)
  if #Unit.allInZone(zone, coalition.side.RED, Group.Category.GROUND) ~= 0 then return end
  if vpa.names.unitCounter.counter >= vpaMax then return end
  if #table.fromiter(Unit.filtered(coalition.side.RED, Group.Category.GROUND)) >= 300 then return end
  local units = Units()
  units:addType('Infantry AK', 7)
  units:addType('Paratrooper RPG-16', 1)
  units:setGoodSkill()
  units:setRandomTransportable(false)
  vpa.names:applyTo(units)
  units:formSquareInZones(zone, vpa.zone)
  units:spawn(country.id.RUSSIA, Group.Category.GROUND)
end

-- Directs all VPA groups to attack blue units within the given zone, the Chu
-- Pong mountain area.
--
-- Be careful to avoid lagging. However, at the same time, take care to balance
-- the task assignments. Otherwise, favoured locations will spawn first when the
-- spawn zone becomes free. Instead pick one at random.
function vpa.attackInZone(zone)
  if #coalition.getPlayers(coalition.side.RED) > 0 then return end
  local units = Unit.allInZone(zone, coalition.side.BLUE, Group.Category.GROUND)
  for unit in Unit.unitsInZone(zone, coalition.side.BLUE, Group.Category.HELICOPTER) do
    table.insert(units, unit)
  end
  local filteredGroups = Group.filtered(coalition.side.RED, Group.Category.GROUND, function(group)
    return vpa.names:includesGroup(group)
  end)
  if #units == 0 then
    for group in filteredGroups do
      group:getController():resetTask()
    end
  else
    local groups = table.fromiter(filteredGroups)
    local group = groups[math.random(#groups)]
    group:setTurningToUnitsTask(units, AI.Task.VehicleFormation.OFF_ROAD, 10)
  end
end
